### DISCUSSION TITLE
How to set experimental log options programatically

### DISCUSSION THREAD

**Discussion by megla-tlanghorst** [open-telemetry/opentelemetry-dotnet#5082] [General]
We are using OTLP logging currently and it's working really good so far, but we also want to include exceptions and stacktraces into the logs. I've seen there's support for that now, but it's gated behind `ExperimentalOptions.EmitLogExceptionAttributes` which can be set by environment variable, but I need to set it programmatically, because we control these options from a central spot programatically.

So far the only viable thing I see there would be to set it from within the process:
```csharp
Environment.SetEnvironmentVariable("OTEL_DOTNET_EXPERIMENTAL_EMIT_EXCEPTION_LOG_ATTRIBUTES", "true", EnvironmentVariableTarget.Process);
```

Is there any other way to set it from the dependency injection?
For the normal options for example we use something like this:
```csharp
services.AddOptions<OpenTelemetryLoggerOptions>()
    .Configure<IOptions<TelemetryOptions>>(
        (options, monitoringOptions) =>
        {
            var config = monitoringOptions.Value;
            options.AddOtlpExporter(
                exporterOptions =>
                {
                    // [...]
                });
        });
```

**Comment by vishweshbankwar** [comment]
@megla-tlanghorst Experimental options can be set from IConfig. Note that this feature was recently added and is not yet available, it is expected to be released with 1.7.0 preview version in a week or so. As an e.g.

```csharp
using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.Services.AddSingleton<IConfiguration>(config);

    builder.AddOpenTelemetry(logging => logging.AddOtlpExporter());
});

```

**Reply by CodeBlanch** [reply]
@megla-tlanghorst 

What @vishweshbankwar has there will work fine for `LoggerFactory.Create` but just in case you were looking for something at the host level here is how you could do that:

```csharp
var appBuilder = WebApplication.CreateBuilder(args);

appBuilder.Configuration.AddInMemoryCollection(
    new Dictionary<string, string?>
    {
        ["OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES"] = "true",
    });

appBuilder.Logging.AddOpenTelemetry(
   options => options.AddOtlpExporter());
```

PS: This won't work with _anything_ pushed to NuGet (currently latest is 1.7.0-alpha.1) but it will work starting with the next release. It needs #4916 to work.

**Reply by megla-tlanghorst** [reply]
But both approaches would modify the global `IConfiguration`, right? Would prefer to not touch that as it's common code for many applications and I can't control what they are doing with their configuration and don't want to break stuff. Would be cool if it could be configured like the `OpenTelemetryLoggerOptions`, right from the dependency injection (also supporting named options like the others).

So something like:
```csharp
services.Configure<BatchExportLogRecordProcessorExperimentalOptions>(opts => {
    // ...
})
```

https://github.com/open-telemetry/opentelemetry-dotnet/pull/4916 seems to already do it for non-experimental settings, at lest according to the description.
